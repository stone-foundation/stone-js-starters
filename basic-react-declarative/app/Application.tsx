import { ReactNode } from "react"
import { Browser } from "@stone-js/browser-adapter"
import { ILogger, LogLevel, StoneApp } from "@stone-js/core"
import { IComponentEventHandler, ReactIncomingEvent, RenderContext, UseReact } from "@stone-js/use-react"

/**
 * Application
 * 
 * This is the main application entry point.
 * 
 * @UseReact() is used to enable the React.
 * @Browser() is used to enable the Browser adapter.
 * @StoneApp() is used to enable the Stone application, it is required.
 */
@Browser()
@UseReact()
@StoneApp({ logger: { level: LogLevel.INFO } })
export class Application implements IComponentEventHandler<ReactIncomingEvent> {
  /**
   * Logger is a service for logging.
  */
  private readonly logger: ILogger

  /**
   * Create a new instance of Application
   * At this point, all the dependencies are resolved and injected.
   * You can access the container and all the services.
   * 
   * @param logger - Logger service.
  */
  constructor ({ logger }: { logger: ILogger }) {
    this.logger = logger
  }

  /**
   * Handle incoming events
   * 
   * @param event Incoming event
   * @returns response
  */
  handle(event: ReactIncomingEvent): ResponseData {
    // Get the name from the event
    const message = `Hello ${event.get<string>('name', 'World')}!`

    // Log a message
    this.logger.info(message)

    // Return a JSON response
    return { message }
  }

  /**
   * Render the component.
   * 
   * @returns The rendered component.
   */
  render ({ data }: RenderContext<ResponseData>): ReactNode {
    return (
      <section className="container">
        <h1 className="h1 text-center mt-64">{data?.message}</h1>
      </section>
    )
  }
}

/**
 * Response data
 */
export interface ResponseData {
  message: string
}
