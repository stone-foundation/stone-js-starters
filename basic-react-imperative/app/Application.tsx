import { JSX } from 'react'
import { browserAdapterBlueprint } from '@stone-js/browser-adapter'
import { defineConfig, IBlueprint, ILogger, isNotEmpty, Promiseable } from '@stone-js/core'
import { defineCommand, FactoryCommandHandler, NODE_CONSOLE_PLATFORM, nodeConsoleAdapterBlueprint } from '@stone-js/node-cli-adapter'
import { defineStoneReactApp, HeadContext, IPage, PageHeadContext, PageRenderContext, ReactIncomingEvent } from '@stone-js/use-react'

/**
 * Create an handler using the factory handler.
 */
export const FactoryHandler = ({ logger }: AppOptions): IPage<ReactIncomingEvent> => {
  return {
    handle (event: ReactIncomingEvent): ResponseData {
      // Get the name from the event
      const message = `Hello ${String(event.get<string>('name', 'World'))}!`

      // Log a message
      logger.info(message)

      // Return a JSON response
      return { message }
    },

    /**
     * Set the page head tags, like title, meta, link, script, style.
     *
     * @returns The head context.
     */
    head ({ event }: PageHeadContext): Promiseable<HeadContext> {
      return {
        title: `Hello ${String(event.get<string>('name', 'World'))}!`,
        description: 'This is a simple example of a React application using Stone.js',
        metas: [
          { name: 'author', content: 'Stone.js' },
          { name: 'keywords', content: 'stonejs,js,react,example' }
        ]
      }
    },

    /**
     * Render the component.
     *
     * @returns The rendered component.
     */
    render ({ data }: PageRenderContext<ResponseData>): JSX.Element {
      return (
        <section className='container'>
          <h1 className='h1 text-center mt-64'>{data?.message}</h1>
        </section>
      )
    }
  }
}

/**
 * Stone-React application.
 */
export const Application = defineStoneReactApp(
  FactoryHandler,
  { debug: true, isFactory: true },
  [browserAdapterBlueprint, nodeConsoleAdapterBlueprint]
)

/**
 * Application configuration.
 */
export const AppConfig = defineConfig({
  afterConfigure (blueprint: IBlueprint) {
    if (
      isNotEmpty<FactoryCommandHandler>(FactoryHandler) &&
      blueprint.is('stone.adapter.platform', NODE_CONSOLE_PLATFORM)
    ) {
      blueprint.set(defineCommand(FactoryHandler, { name: '*', isFactory: true }))
    }
  }
})

/**
 * Application options.
 */
interface AppOptions {
  logger: ILogger
}

/**
 * Response data
 */
interface ResponseData {
  message: string
}
