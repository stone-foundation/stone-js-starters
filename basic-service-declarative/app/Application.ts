export { stoneApp } from "@stone-js/core";
import { NodeHttp } from "@stone-js/node-http-adapter"
import { IncomingHttpEvent } from "@stone-js/http-core"
import { IApplicationHandler, ILogger, LogLevel, StoneApp } from "@stone-js/core"

/**
 * Application options.
 */
interface AppOptions {
  logger: ILogger
}

/**
 * Application
 * 
 * This is the main application entry point.
 * 
 * @StoneApp() is used to enable the Stone application, it is mandatory.
 * @NodeHttp() is used to enable the Node HTTP adapter.
 * 
 * You have access to the Stone's application lifecycle hooks.
 */
@NodeHttp()
@StoneApp({ logger: { level: LogLevel.INFO } })
export class Application implements IApplicationHandler<IncomingHttpEvent> {
  /**
   * Logger is a service for logging.
  */
  private readonly logger: ILogger

  /**
   * Create a new instance of Application
   * At this point, all the dependencies are resolved and injected.
   * You can access the container and all the services.
   * 
   * @param logger - Logger service
  */
  constructor ({ logger }: AppOptions) {
    this.logger = logger
  }

  /**
   * Handle incoming HTTP events
   * 
   * @param event Incoming HTTP event
   * @returns Outgoing HTTP response
  */
  handle(event: IncomingHttpEvent): { message: string } {
    // Get the name from the event
    const message = `Hello ${event.get<string>('name', 'World')}`

    // Log a message
    this.logger.info(message)

    // Return a JSON response
    return { message }
  }
}