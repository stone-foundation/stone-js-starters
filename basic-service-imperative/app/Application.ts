import { NODE_HTTP_PLATFORM, nodeHttpAdapterBlueprint } from "@stone-js/node-http-adapter"
import { defineCommand, NODE_CONSOLE_PLATFORM, nodeConsoleAdapterBlueprint } from "@stone-js/node-cli-adapter"
import { defineStoneApp, defineConfig, IBlueprint, ILogger, IncomingEvent, FunctionalEventHandler } from "@stone-js/core"

/**
 * Create an handler using the factory handler.
 */
export const Application = ({ logger }: AppOptions): FunctionalEventHandler<IncomingEvent> => {
  return (event: IncomingEvent): ResponseData => {
    // Get the name from the event
    const message = `Hello ${event.get<string>('name', 'World')}!`

    // Log a message
    logger.info(message)

    // Return a JSON response
    return { message }
  }
}

/**
 * Application factory.
 *
 * @param options - The application options.
 * @returns A function that handles incoming events and returns a response.
 */
export const MyStoneApp = defineStoneApp(
  Application,
  {
    debug: true,
    isFactory: true,
    adapter: { platform: NODE_HTTP_PLATFORM }
  },
  [nodeHttpAdapterBlueprint, nodeConsoleAdapterBlueprint],
)

/**
 * Application configuration.
 */
export const AppConfig = defineConfig({
  afterConfigure (blueprint: IBlueprint) {
    if (blueprint.is('stone.adapter.platform', NODE_CONSOLE_PLATFORM)) {
      blueprint.set(defineCommand(Application, { name: '*', isFactory: true }))
    }
  }
})

/**
 * Application options.
 */
interface AppOptions {
  logger: ILogger
}

/**
 * Response data
 */
export interface ResponseData {
  message: string
}
