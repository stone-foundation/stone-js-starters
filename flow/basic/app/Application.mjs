import { HttpResponse } from '@stone-js/http'
import { StoneApp } from '@stone-js/core/decorators'
import { NodeHttpAdapter } from '@stone-js/adapters/decorators'

@StoneApp()
@NodeHttpAdapter({ default: true })
export class Application {
  /**
   * Create an Application instance.
   *
   * @param {Container} container - Service container. For more details, see: https://www.npmjs.com/package/@stone-js/service-container.
   */
  constructor (container) {
    this.container = container
  }

  /**
   * OnInit Hook
   * Hook that runs once and only once when the application starts.
   * Useful for initialization and configuration tasks at startup.
   * 
   * @returns
   */
  static onInit () {
    console.log('I am executed once and for all!')
  }

  /**
   * BeforeHandle Hook
   * Hook that runs before each event and just before the action handler.
   * Useful for initialization and configuration that needs to be done before each event.
   * 
   * @returns
   */
  beforeHandle () {
    console.log('I am executed before each event!')
  }

  /**
   * Event Handler: The Main Entry Point for Your Application
   * Add your custom application logic here to process the event.
   * 
   * @param   {IncomingHttpEvent} event - Stone IncomingEvent. For more details, see: https://github.com/stonemjs/common/blob/main/src/IncomingEvent.mjs.
   * @returns {OutgoingHttpResponse} Stone OutgoingHttpResponse. For more details, see: https://github.com/stonemjs/http/blob/main/src/HttpResponse.mjs.
   */
  async handle (event) {
    console.log('I am here to handle events:', event.method, event.path)
    return HttpResponse.json({ message: 'Hello World' })
  }

  /**
   * Terminate Hook
   * This hook runs after each response has been sent to the user.
   * Add your custom logic here to handle cleanup and finalization tasks.
   * 
   * @returns
   */
  onTerminate () {
    console.log('I am executed after each response is sent!')
  }
}