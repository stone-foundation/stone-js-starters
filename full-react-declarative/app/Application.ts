import { getString } from "@stone-js/env";
import { Routing } from "@stone-js/router";
import { Browser } from "@stone-js/browser-adapter";
import { NodeHttp } from "@stone-js/node-http-adapter";
import { NodeConsole } from "@stone-js/node-cli-adapter";
import { ReactIncomingEvent, ReactOutgoingResponse, Hook, UseReact } from "@stone-js/use-react";
import { IAdapterHook, IBlueprint, IContainer, IKernelHook, ILogger, LogLevel, Promiseable, StoneApp } from "@stone-js/core";

/**
 * Application
 * 
 * This is the main application entry point.
 * 
 * @StoneApp() is used to enable the Stone application, it is mandatory.
 * @Routing() is used to enable the routing feature.
 * @NodeHttp() is used to enable the Node HTTP adapter.
 * @NodeConsole() is used to enable the Node Console adapter.
 * @NodeConsole() requires the incoming event type.
 * BodyEventMiddleware is used to parse the incoming event body.
 * BodyEventMiddleware is imported because it is not installed by default.
 * 
 * You have access to the Stone's application lifecycle hooks.
 * You can also have access to them in the ServiceProviders.
 * They are optional and you can remove them if you don't need them.
 */
@Routing()
@Browser()
@UseReact()
@NodeConsole()
@StoneApp({ name: 'My Stone', logger: { level: LogLevel.INFO } })
@NodeHttp({ default: true, url: getString('BASE_URL', 'http://localhost:3100') })
export class Application implements IAdapterHook, IKernelHook<ReactIncomingEvent, ReactOutgoingResponse> {
  /**
   * Start the application
   * Run at each cold start.
   * At this point, you only have access to the blueprint.
   * Because the container is not yet created.
   */
  @Hook('onStart')
  onStart({ blueprint }: { blueprint: IBlueprint }): void {
    const AppName = blueprint.get<string>('stone.name')
    console.log(`${AppName}'s Application is starting`)
  }

  /**
   * Prepare the application
   * Run at each warm start in server context.
   * And at each cold start in browser context.
   */
  @Hook('onInit')
  onInit({ logger }: { logger: ILogger }): Promiseable<void> {
    logger.info('Application is initializing')
  }

  /**
   * Before handle the incoming event
   */
  @Hook('onHandlingEvent')
  onHandlingEvent({ logger }: { logger: ILogger }): Promiseable<void> {
    logger.info('Before handle incoming event')
  }

  /**
   * Before prepare the react page
   */
  @Hook('onPreparingPage')
  onPreparingPage({ container }: { container: IContainer }): Promiseable<void> {
    container.make<ILogger>('logger').info('Before prepare the react page')
  }

  /**
   * After the incoming event has been processed
   */
  @Hook('onProcessingKernelMiddleware')
  onProcessingKernelMiddleware({ pipe }: { pipe: any }): Promiseable<void> {
    console.info('Kernel middleware is processing', pipe.module.name)
  }

  /**
   * After the platform event has been sent to the user
   */
  @Hook('onTerminate')
  onTerminate ({ logger }: { logger: ILogger }): Promiseable<void> {
    logger.info('The event has been sent to the user')
  }
}
