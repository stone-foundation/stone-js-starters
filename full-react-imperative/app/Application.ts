import { routerBlueprint } from "@stone-js/router";
import { browserAdapterBlueprint } from "@stone-js/browser-adapter";
import { nodeHttpAdapterBlueprint } from "@stone-js/node-http-adapter";
import { nodeConsoleAdapterBlueprint } from "@stone-js/node-cli-adapter";
import { defineStoneReactApp, ReactLifecycleHookType } from "@stone-js/use-react";
import { defineHookListeners, IBlueprint, IContainer, ILogger, LogLevel, Promiseable } from "@stone-js/core";

/**
 * Start the application
 * Run at each cold start.
 * At this point, you only have access to the blueprint.
 * Because the container is not yet created.
 */
export function onStart({ blueprint }: { blueprint: IBlueprint }): void {
  const AppName = blueprint.get<string>('stone.name')
  console.log(`${AppName}'s Application is starting`)
}

/**
 * Prepare the application
 * Run at each warm start in server context.
 * And at each cold start in browser context.
 */
export function onInit({ logger }: { logger: ILogger }): Promiseable<void> {
  logger.info('Application is initializing')
}

/**
 * Before handle the incoming event
 */
export function onHandlingEvent({ logger }: { logger: ILogger }): Promiseable<void> {
  logger.info('Before handle incoming event')
}

/**
 * Before prepare the react page
 */
export function onPreparingPage({ container }: { container: IContainer }): Promiseable<void> {
  container.make<ILogger>('logger').info('Before prepare the react page')
}

/**
 * After the incoming event has been processed
 */
export function onProcessingKernelMiddleware({ pipe }: { pipe: any }): Promiseable<void> {
  console.info('Kernel middleware is processing', pipe.module.name)
}

/**
 * After the platform event has been sent to the user
 */
export function onTerminate ({ logger }: { logger: ILogger }): Promiseable<void> {
  logger.info('The event has been sent to the user')
}

/**
 * Stone-React Application
 */
export const MyStoneReactApp = defineStoneReactApp(
  { name: 'My Stone', logger: { level: LogLevel.DEBUG } },
  [
    routerBlueprint,
    browserAdapterBlueprint,
    nodeHttpAdapterBlueprint,
    nodeConsoleAdapterBlueprint
  ]
)

/**
 * Hooks Listeners Blueprint
 */
export const HooksListenersBlueprint = defineHookListeners<ReactLifecycleHookType>({
  onInit: [onInit],
  onStart: [onStart],
  onTerminate: [onTerminate],
  onPreparingPage: [onPreparingPage],
  onHandlingEvent: [onHandlingEvent],
  onProcessingKernelMiddleware: [onProcessingKernelMiddleware]
})
