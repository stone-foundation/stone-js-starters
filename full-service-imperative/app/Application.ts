import {
  IncomingHttpEvent,
  OutgoingHttpResponse,
  IncomingHttpEventOptions
} from '@stone-js/http-core'
import {
  isEmpty,
  LogLevel,
  IContainer,
  IBlueprint,
  defineConfig,
  AdapterConfig,
  defineStoneApp,
  AdapterContext,
  BlueprintContext,
  defineHookListener,
  AdapterEventBuilderType,
  AdapterHookListenerContext
} from '@stone-js/core'
import { getString } from '@stone-js/env'
import { routerBlueprint } from '@stone-js/router'
import { PipelineHookContext } from '@stone-js/pipeline'
import { nodeConsoleAdapterBlueprint } from '@stone-js/node-cli-adapter'
import { MetaBodyEventMiddleware, MetaFilesEventMiddleware, NODE_HTTP_PLATFORM, nodeHttpAdapterBlueprint } from '@stone-js/node-http-adapter'

/**
 * Run before the blueprint is prepared
 *
 * @param context - The blueprint context
 */
export const onPreparingBlueprint = defineHookListener(({ blueprint, modules }: BlueprintContext): void => {
  blueprint.set('tmpBlueprintTimer', Date.now())
  console.log(`Preparing blueprint by introspecting ${modules.length} modules...`)
}, { name: 'onPreparingBlueprint' })

/**
 * Run before each blueprint middleware is processed
 *
 * @param context - The pipeline context
 */
export const onProcessingBlueprintMiddleware = defineHookListener(({ passable, pipe }: PipelineHookContext<BlueprintContext, IBlueprint>): void => {
  passable.blueprint.set(`tmpBlueprintMiddleware.${String(pipe.module.name)}`, Date.now())
}, { name: 'onProcessingBlueprintMiddleware' })

/**
 * Run after each blueprint middleware is processed
 *
 * @param context - The pipeline context
 */
export const onBlueprintMiddlewareProcessed = defineHookListener(({ passable, pipe }: PipelineHookContext<BlueprintContext, IBlueprint>): void => {
  const name = `tmpBlueprintMiddleware.${String(pipe.module.name)}`
  const start = passable.blueprint.get(name, 0)
  const Type = pipe.isClass !== undefined ? 'Class' : pipe.isFactory !== undefined ? 'Factory' : pipe.isAlias !== undefined ? 'Alias' : 'Callable'
  const value = {
    Middleware: pipe.module.name,
    Type,
    Priority: pipe.priority,
    'Execution Time': `${Date.now() - start}ms`
  }

  passable.blueprint.set(name, value)
}, { name: 'onBlueprintMiddlewareProcessed' })

/**
 * Run after the blueprint is prepared
 *
 * @param context - The blueprint context
 */
export const onBlueprintPrepared = defineHookListener(({ blueprint }: BlueprintContext): void => {
  const timer = Date.now() - blueprint.get('tmpBlueprintTimer', 0)
  const middleware = Object.values(blueprint.get('tmpBlueprintMiddleware', {}))
  const message = `Blueprint prepared in ${timer}ms, with ${middleware.length} middleware`

  console.table(middleware)
  console.log(message)
}, { name: 'onBlueprintPrepared' })

/**
 * Run before the application starts
 *
 * @param context - The adapter context
 */
export const onStart = defineHookListener(({ blueprint, context }: AdapterHookListenerContext): void => {
  console.log('Starting application...')
  console.log('Context isEmpty?', isEmpty(context))
  console.log('Blueprint keys:', Object.keys(blueprint.all()).length)
}, { name: 'onStart' })

/**
 * Run before each adapter middleware is processed
 *
 * @param context - The pipeline context
 */
export const onProcessingAdapterMiddleware = defineHookListener(({ passable, pipe, pipes }: PipelineHookContext<
AdapterContext<unknown, unknown, unknown, IncomingHttpEvent, IncomingHttpEventOptions, OutgoingHttpResponse>,
AdapterEventBuilderType<any>
>): void => {
  console.log('Processing adapter middleware...')
  console.log('Middleware count:', pipes.length)
  console.log('Middleware name:', pipe.name)
  console.log('IncomingEvent isEmpty?', isEmpty(passable.incomingEvent))
}, { name: 'onProcessingAdapterMiddleware' })

/**
 * Run after each adapter middleware is processed
 *
 * @param context - The pipeline context
 */
export const onAdapterMiddlewareProcessed = defineHookListener(({ passable, pipe, pipes }: PipelineHookContext<
AdapterContext<unknown, unknown, unknown, IncomingHttpEvent, IncomingHttpEventOptions, OutgoingHttpResponse>,
AdapterEventBuilderType<any>
>): void => {
  console.log('Adapter middleware processed...')
  console.log('Middleware count:', pipes.length)
  console.log('Middleware name:', pipe.name)
  console.log('IncomingEvent isEmpty?', isEmpty(passable.incomingEvent))
}, { name: 'onAdapterMiddlewareProcessed' })

/**
 * Run before building each incoming event
 *
 * @param context - The adapter context
 */
export const onBuildingIncomingEvent = defineHookListener(({ context }: AdapterHookListenerContext<
AdapterContext<unknown, unknown, unknown, IncomingHttpEvent, IncomingHttpEventOptions, OutgoingHttpResponse>
>): void => {
  console.log('Building incoming event...')
  console.log('IncomingEvent isEmpty?', isEmpty(context?.incomingEvent))
}, { name: 'onBuildingIncomingEvent' })

/**
 * Run before handling each adapter error
 *
 * @param context - The adapter context
 */
export const onHandlingAdapterError = defineHookListener(({ error }: AdapterHookListenerContext<
AdapterContext<unknown, unknown, unknown, IncomingHttpEvent, IncomingHttpEventOptions, OutgoingHttpResponse>
>): void => {
  console.error('Handling adapter error...')
  console.error('Error:', error)
}, { name: 'onHandlingAdapterError' })

/**
 * Run before building each outgoing response
 *
 * @param context - The adapter context
 */
export const onBuildingRawResponse = defineHookListener(({ context }: AdapterHookListenerContext<
AdapterContext<unknown, unknown, unknown, IncomingHttpEvent, IncomingHttpEventOptions, OutgoingHttpResponse>
>): void => {
  console.log('Building raw response...')
  console.log('IncomingEvent isEmpty?', isEmpty(context?.incomingEvent))
  console.log('OutgoingResponse isEmpty?', isEmpty(context?.outgoingResponse))
}, { name: 'onBuildingRawResponse' })

/**
 * Run just before the application stops
 *
 * @param context - The adapter context
 */
export const onStop = defineHookListener(({ blueprint, context }: AdapterHookListenerContext): void => {
  console.log('Stopping application...')
  console.log('Context isEmpty?', isEmpty(context))
  console.log('Blueprint keys:', Object.keys(blueprint.all()).length)
}, { name: 'onStop' })

/**
 * Run after the application is initialized
 * The container is available at this point.
 * All services are registered and ready to be used.
 *
 * @param container - The container
 */
export const onInit = defineHookListener((container: IContainer): void => {
  console.log('Initializing application...')
  console.log('Container services count:', container.getBindings().size)
}, { name: 'onInit' })

/**
 * Run before handling each event
 *
 * @param container - The container
 */
export const onHandlingEvent = defineHookListener((container: IContainer): void => {
  try {
    console.log('Handling event...')
    container.make('event')
  } catch (error: any) {
    console.log('The incoming event is not yet bound at this stage', error.message)
  }
}, { name: 'onHandlingEvent' })

/**
 * Run before processing each kernel middleware
 *
 * @param context - The pipeline context
 */
export const onProcessingKernelMiddleware = defineHookListener((
  { passable, pipe, pipes }: PipelineHookContext<IncomingHttpEvent, OutgoingHttpResponse>
): void => {
  console.log('Processing kernel middleware...')
  console.log('Middleware count:', pipes.length)
  console.log('Middleware name:', pipe.name)
  console.log('Incoming Source name', passable.source.platform)
}, { name: 'onProcessingKernelMiddleware' })

/**
 * Run after processing each kernel middleware
 *
 * @param context - The pipeline context
 */
export const onKernelMiddlewareProcessed = defineHookListener((
  { passable, pipe, pipes }: PipelineHookContext<IncomingHttpEvent, OutgoingHttpResponse>
): void => {
  console.log('Kernel middleware processed...')
  console.log('Middleware count:', pipes.length)
  console.log('Middleware name:', pipe.name)
  console.log('Incoming Source name', passable.source.platform)
}, { name: 'onTerminate' })

/**
 * Run before executing each event handler
 *
 * @param container - The container
 */
export const onExecutingEventHandler = defineHookListener(({ event }: { event: IncomingHttpEvent }): void => {
  console.log('Executing event handler...')
  console.log('The incoming event is yet bound at this stage', event.source.platform)
}, { name: 'onExecutingEventHandler' })

/**
 * Run before handling each error
 *
 * @param container - The container
 */
export const onExecutingErrorHandler = defineHookListener(({ error }: { error: any }): void => {
  console.log('Executing error handler...')
  console.log('Error', error)
}, { name: 'onExecutingErrorHandler' })

/**
 * Run before preparing each response
 *
 * @param container - The container
 */
export const onPreparingResponse = defineHookListener(({ response }: { response: OutgoingHttpResponse }): void => {
  console.log('Preparing response...')
  console.log('Response headers', response.headers)
}, { name: 'onPreparingResponse' })

/**
 * Run after preparing each response
 *
 * @param container - The container
 */
export const onResponsePrepared = defineHookListener(({ response }: { response: OutgoingHttpResponse }): void => {
  console.log('Response prepared...')
  console.log('Response headers', response.headers)
}, { name: 'onResponsePrepared' })

/**
 * Run after handling each event
 *
 * @param context - The pipeline context
 */
export const onEventHandled = defineHookListener((
  { event, response }: { event: IncomingHttpEvent, response: OutgoingHttpResponse }
): void => {
  console.log('Event handled...')
  console.log('Event headers', event.headers)
  console.log('Response headers', response.headers)
}, { name: 'onEventHandled' })

/**
 * Run after executing each event handler
 *
 * @param container - The container
 */
export const onTerminate = defineHookListener((container: IContainer): void => {
  console.log('Terminating event handling...')
  console.log('Is IncomingEvent exist', container.bound('event'))
  console.log('Is OutgoingResponse exist', container.bound('response'))
}, { name: 'onTerminate' })

/**
 * Define Stone Application.
 */
export const Application = defineStoneApp(
  { name: 'MyApp', logger: { level: LogLevel.INFO } },
  [
    routerBlueprint,
    nodeHttpAdapterBlueprint,
    nodeConsoleAdapterBlueprint
  ]
)

/**
 * Application configuration.
 */
export const AppConfig = defineConfig({
  configure (blueprint: IBlueprint) {
    blueprint
      .get<AdapterConfig[]>('stone.adapters', [])
      .forEach((adapter: AdapterConfig) => {
        if (adapter.platform === NODE_HTTP_PLATFORM) {
          adapter.default = true
        }
      })
  },
  afterConfigure (blueprint: IBlueprint) {
    if (blueprint.is('stone.adapter.platform', NODE_HTTP_PLATFORM)) {
      blueprint
        .set('stone.adapter.url', getString('BASE_URL', 'http://localhost:8080'))
        .add('stone.adapter.middleware', [MetaBodyEventMiddleware, MetaFilesEventMiddleware])
    }
  }
})
