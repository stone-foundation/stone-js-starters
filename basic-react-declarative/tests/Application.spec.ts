import { ILogger } from '@stone-js/core'
import { Application } from '../app/Application'
import { renderToString } from 'react-dom/server'
import { ReactIncomingEvent } from '@stone-js/use-react'

// We must mock decorators to lighten the test environment
vi.mock('@stone-js/core', async (importOriginal) => {
  const actual: any = await importOriginal()
  return {
    ...actual,
    StoneApp: vi.fn(() => vi.fn()),
  }
})

// We must mock decorators to lighten the test environment
vi.mock('@stone-js/node-cli-adapter', async (importOriginal) => {
  const actual: any = await importOriginal()
  return {
    ...actual,
    NodeConsole: vi.fn(() => vi.fn()),
  }
})

// We must mock decorators to lighten the test environment
vi.mock('@stone-js/browser-adapter', async (importOriginal) => {
  const actual: any = await importOriginal()
  return {
    ...actual,
    Browser: vi.fn(() => vi.fn()),
  }
})

// We must mock decorators to lighten the test environment
vi.mock('@stone-js/use-react', async (importOriginal) => {
  const actual: any = await importOriginal()
  return {
    ...actual,
    UseReact: vi.fn(() => vi.fn()),
  }
})

describe('Application', () => {
  let app: Application
  let mockedLogger: ILogger

  beforeEach(() => {
    mockedLogger = {
      info: vi.fn(),
    } as unknown as ILogger

    app = new Application({ logger: mockedLogger })
  })

  it('should create an application instance', () => {
    // Assert
    expect(app).toBeInstanceOf(Application)
  })

  it('get head values', async () => {
    // Arrange
    const expectedMessage = 'Hello World!'
    const event = { get: () => 'World' } as unknown as ReactIncomingEvent

    // Act
    const head = await app.head({ event })

    // Assert
    expect(head).toHaveProperty('metas')
    expect(head.title).toBe(expectedMessage)
    expect(head).toHaveProperty('description')
  })

  it('should handle incoming events', () => {
    // Arrange
    const expectedMessage = 'Hello World!'
    const event = { get: () => 'World' } as unknown as ReactIncomingEvent

    // Act
    const response = app.handle(event) as { message: string }

    // Assert
    expect(response.message).toBe(expectedMessage)
    expect(mockedLogger.info).toHaveBeenCalledWith(expectedMessage)
  })

  it('should render component', () => {
    // Arrange
    const message = 'Hello World!'

    // Act
    const response = renderToString(app.render({ data: { message } } as any))

    // Assert
    expect(response).toMatchSnapshot()
  })
})