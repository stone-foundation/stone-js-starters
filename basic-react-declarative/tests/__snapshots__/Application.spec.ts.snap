// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Application > should render component 1`] = `"<section class="container"><h1 class="h1 text-center mt-64">Hello World!</h1></section>"`;
